factor/grace/classes/factor.php

```diff
@@ -0,0 +1,158 @@
+<?php
+// This file is part of Moodle - http://moodle.org/
+//
+// Moodle is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// Moodle is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.
+
+/**
+ * Grace period factor class.
+ *
+ * @package     tool_mfa
+ * @author      Peter Burnett <peterburnett@catalyst-au.net>
+ * @copyright   Catalyst IT
+ * @license     http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+
+namespace factor_grace;
+
+defined('MOODLE_INTERNAL') || die();
+
+use tool_mfa\local\factor\object_factor_base;
+
+class factor extends object_factor_base {
+
+    /**
+     * Grace Factor implementation.
+     * This factor needs no user setup, return true.
+     *
+     * {@inheritDoc}
+     */
+    public function setup_user_factor($data) {
+        return true;
+    }
+
+    /**
+     * Grace Factor implementation.
+     * This factor is a singleton, return single instance.
+     *
+     * {@inheritDoc}
+     */
+    public function get_all_user_factors() {
+        global $DB, $USER;
+
+        if (!$DB->record_exists('factor_grace', array('userid' => $USER->id))) {
+            $DB->insert_record('factor_grace', array(
+                'userid' => $USER->id,
+                'ip' => $USER->lastip,
+                'timecreated' => time(),
+            ));
+        }
+        $factorrecord = $DB->get_record('factor_grace', array('userid' => $USER->id));
+
+        $factor = (object) array(
+            'id' => 1,
+            'name' => $this->name,
+            'devicename' => '-',
+            'timecreated' => $factorrecord->timecreated,
+            'createdfromip' => $factorrecord->timecreated,
+            'lastverified' => '-',
+            'revoked' => '-'
+        );
+
+        return [$factor];
+    }
+
+    /**
+     * Grace Factor implementation.
+     * Factor has no input.
+     *
+     * {@inheritDoc}
+     */
+    public function has_input() {
+        return false;
+    }
+
+    /**
+     * Grace Factor implementation.
+     * Checks the user login time against their first login after MFA activation.
+     *
+     * {@inheritDoc}
+     */
+    public function get_state() {
+        // Check if user already has a record, if not, create one at current time.
+        global $USER, $DB;
+
+        if (!$DB->record_exists('factor_grace', array('userid' => $USER->id))) {
```

Comment:
can't we just call get_all_user_factors() here and up the first and only record in that array?